### PR TITLE
vim-patch:9.0.0285: it is not easy to change the command line from a plugin

### PR DESCRIPTION
**Pull request by Shougo**

### CODE REVIEW

**Comment by Shougo** [comment]
I may include the patch.

https://github.com/vim/vim/pull/10995

**Comment by zeertzjq** [comment]
That patch requires several functions that are not in Nvim yet.

File: runtime/doc/builtin.txt [hunk 1]
```diff
@@ -6912,6 +6917,16 @@ setcharsearch({dict})					*setcharsearch()*
...
 
+setcmdline({str} [, {pos}])					*setcmdline()*
+		Set the command line to {str} and set the cursor position to
+		{pos}.
+		If {pos} is omitted, the cursor is positioned after the text.
+		Returns 0 when successful, 1 when not editing the command
+		line.
+
+		Can also be used as a |method|: >
+			SetCmdline()->setcmdline()
```
Comment (zeertzjq) [inline runtime/doc/builtin.txt:6928]:
```suggestion
			GetText()->setcmdline()
```

File: src/nvim/eval/funcs.c [hunk 2]
```diff
@@ -7619,6 +7619,31 @@ static void f_setcharsearch(typval_T *argvars, typval_T *rettv, EvalFuncData fpt
...
+    if (error) {
+      return;
+    }
+    if (pos < 0) {
+      emsg(_(e_positive));
+      return;
+    }
+  }
+
+  rettv->vval.v_number = set_cmdline_str(tv_get_string_chk(&argvars[0]), pos);
```
Comment (zeertzjq) [inline src/nvim/eval/funcs.c:7644]: `VAR_STRING` is already checked about, so just use `v_string` here.
```suggestion
  rettv->vval.v_number = set_cmdline_str(argvars[0].vval.v_string, pos);
```

File: src/nvim/ex_getln.c [hunk 3]
```diff
@@ -3991,6 +3967,34 @@ int get_cmdline_screen_pos(void)
   return p->cmdspos;
 }
 
+// Set the command line str to "str".
+// Returns 1 when failed, 0 when OK.
+int set_cmdline_str(const char *str, int pos)
+{
+  struct cmdline_info *p = get_ccline_ptr();
```
Comment (zeertzjq) [inline src/nvim/ex_getln.c:3974]:
```suggestion
  CmdlineInfo *p = get_ccline_ptr();
```

File: src/nvim/ex_getln.c [hunk 4]
```diff
@@ -3991,6 +3967,34 @@ int get_cmdline_screen_pos(void)
...
+  if (p == NULL) {
+    return 1;
+  }
+
+  int len = (int)STRLEN(str);
+  realloc_cmdbuff(len + 1);
+  p->cmdlen = len;
+  STRCPY(p->cmdbuff, str);
+
+  p->cmdpos = pos < 0 ? p->cmdlen : pos;
```
Comment (zeertzjq) [inline src/nvim/ex_getln.c:3985]:
```suggestion
  p->cmdpos = pos < 0 || pos > p->cmdlen ? p->cmdlen : pos;
```

**Comment by Shougo** [comment]
Thanks.

**Pull request merged**